ffi/diplomat/src/bidi.rs

```diff
@@ -0,0 +1,181 @@
+        /// object to the paragraph at that index. Returns an error when out of bounds.
+        ///
+        /// This is equivalent to calling `paragraph_at()` on `ICU4XBidiInfo` but doesn't
+        /// create a new object
+        pub fn set_paragraph_in_text(&mut self, n: usize) -> DiplomatResult<(), ()> {
+            let para = self.0.info.paragraphs.get(n);
+            let para = if let Some(para) = para {
+                para
+            } else {
+                return Err(()).into();
+            };
+
+            self.0 = Paragraph::new(self.0.info, para);
+            Ok(()).into()
+        }
+        #[diplomat::rust_link(unicode_bidi::Paragraph::level_at, FnInStruct)]
+        /// The primary direction of this paragraph
+        pub fn direction(&self) -> ICU4XBidiDirection {
+            self.0.direction().into()
+        }
+
+        /// The number of bytes in this paragraph
+        #[diplomat::rust_link(unicode_bidi::ParagraphInfo::len, FnInStruct)]
+        pub fn size(&self) -> usize {
+            self.0.para.len()
+        }
+
+        /// The start index of this paragraph within the source text
+        pub fn range_start(&self) -> usize {
+            self.0.para.range.start
+        }
+
+        /// The end index of this paragraph within the source text
+        pub fn range_end(&self) -> usize {
+            self.0.para.range.end
+        }
+
+        /// Reorder a line based on display order. The ranges are specified relative to the source text and must be contained
+        /// within this paragraph's range.
+        #[diplomat::rust_link(unicode_bidi::Paragraph::level_at, FnInStruct)]
+        pub fn reorder_line(
+            &self,
+            range_start: usize,
+            range_end: usize,
+            out: &mut DiplomatWriteable,
+        ) -> DiplomatResult<(), ()> {
+            if range_start < self.range_start() || range_end > self.range_end() {
+                return Err(()).into();
+            }
+
+            let info = self.0.info;
+            let para = self.0.para;
+
+            let reordered = info.reorder_line(para, range_start..range_end);
+
+            out.write_str(&reordered).map_err(|_| ()).into()
+        }
+
+        /// Get the BIDI level. This integer is conceptually a `unicode_bidi::Level`,
+        /// and can be further inspected using the static methods on this class.
+        #[diplomat::rust_link(unicode_bidi::Paragraph::level_at, FnInStruct)]
+        pub fn level_at(&self, pos: usize) -> u8 {
+            if pos >= self.size() {
+                return 0;
+            }
+
+            self.0.level_at(pos).number()
+        }
+
+        /// Check if a Level returned by level_at is an RTL level.
+        ///
+        /// Invalid levels (numbers greater than 125) will be assumed LTR
+        #[diplomat::rust_link(unicode_bidi::Level::is_rtl, FnInStruct)]
+        pub fn level_is_rtl(level: u8) -> bool {
```

Comment:
Suggestion: should these static functions be on top-level `ICU4XBidi`?

Reply:
Paragraphs are the ones producing levels so I figured it made the most sense for paragraphs to have them. If we add level APIs to ICU4XBidi we can move them up.

I suspect we'll have some iteration here as we work with Skia, for now I just want something that _works_.

Reply:
Thoughts:

1. We want to be in a position that we are adding things, which is a lot easier than moving things.
2. Since Diplomat doesn't support module-level statics, the closest we have is the "main" class for a component as the namespace to hold them. So I think putting them there is a good option even now.